### PR TITLE
kubernetesdiscovery: propagate PortForward template

### PR DESCRIPTION
**Pull request by nicks**
Hello @milas, @landism,

Please review the following commits I made in branch nicks/pf2:

27780ba24c08472638756280492ff64df3b6f46e (2021-07-02 18:34:35 -0400)
kubernetesdiscovery: propagate PortForward template

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
this is mostly adapted from https://github.com/tilt-dev/tilt/blob/master/internal/engine/portforward/subscriber.go

File: internal/controllers/core/kubernetesdiscovery/portforwards.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		if bestPod == nil || isBetterPortForwardPod(&pod, bestPod) {
+			bestPod = &pod
+		}
+	}
+	return bestPod
+}
+
+// Check if podA is better than podB for port-forwarding.
+func isBetterPortForwardPod(podA, podB *v1alpha1.Pod) bool {
+	// A deleting pod is always better than a non-deleting pod.
```
Comment (milas) [inline internal/controllers/core/kubernetesdiscovery/portforwards.go:164]:
```suggestion
	// A non-deleting pod is always better than a deleting pod.
```
I might have not yet had enough coffee, but I think this should be the opposite? That seems to be what the logic is actually doing - comment just caught me off-guard 😛

File: internal/controllers/core/kubernetesdiscovery/reconciler.go [hunk 2]
```diff
@@ -179,6 +189,9 @@ func (w *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 			w.teardown(key)
 			w.cleanupAbandonedNamespaces()
 		}
+
+		// TODO(nick): Delete any orphaned PortForwards or PodLogStreams.
```
Comment (milas) [inline internal/controllers/core/kubernetesdiscovery/reconciler.go:193]: Ahh whoops, I can't believe I spent the time tracking down the lack of automatic GC on owner references in our apiserver and then didn't actually write the logic to handle it manually 🤦

**Review by milas — APPROVED** [review]

File: internal/controllers/core/kubernetesdiscovery/portforwards.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+	errs := []error{}
+	foundDesired := false
+	for _, existingPF := range pfList.Items {
+		matchesPF := pf != nil && existingPF.Name == pf.Name
+		if matchesPF {
+			foundDesired = true
+
+			// If this PortForward is already in the APIServer, make sure it's up-to-date.
+			if apicmp.DeepEqual(pf.Spec, existingPF) &&
+				existingPF.ObjectMeta.Annotations[v1alpha1.AnnotationManifest] ==
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/controllers/core/kubernetesdiscovery/portforwards.go:43]: oooh, i like that idea. done!

File: internal/controllers/core/kubernetesdiscovery/portforwards.go (hunk 1 above)
Comment (nicks) [reply internal/controllers/core/kubernetesdiscovery/portforwards.go:164]: oops, fixed

File: internal/controllers/core/kubernetesdiscovery/reconciler.go (hunk 2 above)
Comment (nicks) [reply internal/controllers/core/kubernetesdiscovery/reconciler.go:193]: :laughing:

**Pull request merged**